### PR TITLE
DOC: Build versioned docs and deploy them to gh-pages

### PR DESCRIPTION
**Pull request by oesteban**
This PR adds three new build steps to our CircleCI workflow.
The first one, called ``build_docs`` generates the documentation
of the project using sphinx.
In the case the commit message contains the tag ``[docs versions]``
(or a few variations thereof), then a readthedocs-style of versioned
documentation is built.
The other two build steps are run only for tag-commits and the master
branch, to deploy the new versioned documentation to GitHub pages.

I've set up the corresponding write-access keys in the repo and
CircleCI, so this first step towards having documentation should fairly
good to go.

### CODE REVIEW

**Comment by oesteban** [comment]
Finally :). The pre-render of the docs is here: https://154-206670456-gh.circle-artifacts.com/0/root/project/docs/_build/no_version_html/html/index.html

This will be very useful to visualize finalized workflows (see https://154-206670456-gh.circle-artifacts.com/0/root/project/docs/_build/no_version_html/html/api/dmriprep.workflows.base.html)

Also, it will be possible to have a nice view of the docstrings and doctest examples we embed in our code (e.g., latest additions by @dPys on https://154-206670456-gh.circle-artifacts.com/0/root/project/docs/_build/no_version_html/html/api/dmriprep.utils.vectors.html)

**Comment by oesteban** [comment]
An additional feature is: by using Nipype's sphinx extension to generate the workflow graph we are exercising the workflow creation function (is a kind of implicit unit test :))

**Comment by dPys** [comment]
These look great!

**Comment by oesteban** [comment]
Yep, I'm eager to see the graph of new workflows in #25

File: docs/installation.rst [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+     $ docker run -it ubuntu bash
+    
+    Share images, automate workflows, and more with a free Docker ID:
+     https://hub.docker.com/
+    
+    For more examples and ideas, visit:
+     https://docs.docker.com/get-started/
+
+After checking your Docker Engine is capable of running Docker images, then go ahead
+and `check out our documentation <https://fmriprep.readthedocs.io/en/stable/docker.html>`__
```
Comment (josephmje) [inline docs/installation.rst]: fix link

File: docs/installation.rst [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+Please make sure you `follow our tips and tricks to run dMRIPrep's Singularity images
+<singularity.html>`_.
+
+
+Manually Prepared Environment (Python 3.5+)
+===========================================
+.. warning::
+
+   This method is not recommended! Please checkout container alternatives
+   such as `Docker <https://fmriprep.readthedocs.io/en/stable/docker.html>`__, or
```
Comment (josephmje) [inline docs/installation.rst]: fix links

File: docs/installation.rst [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+*dMRIPrep* requires some other neuroimaging software tools that are
+not handled by the Python's packaging system (Pypi) used to deploy
+the ``dmriprep`` package:
+
+- FSL_ (version 5.0.9)
+- ANTs_ (version 2.2.0 - NeuroDocker build)
+- AFNI_ (version Debian-16.2.07)
+- `C3D <https://sourceforge.net/projects/c3d/>`_ (version 1.0.0)
+- FreeSurfer_ (version 6.0.1)
+- `bids-validator <https://github.com/bids-standard/bids-validator>`_ (version 1.1.0)
```
Comment (josephmje) [inline docs/installation.rst]:
```suggestion
- `bids-validator <https://github.com/bids-standard/bids-validator>`_ (version 1.2.3)
```

File: dmriprep/workflows/base.py [hunk 4]
```diff
@@ -222,91 +222,90 @@ def init_single_subject_wf(
     Functional preprocessing is performed using a separate workflow for each
```
Comment (arokem) [inline dmriprep/workflows/base.py]: I just noticed that this says "Functional" and "BOLD"

File: docs/usage.rst [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+.. _fs_license:
+
+The FreeSurfer license
+----------------------
+*dMRIPrep* uses FreeSurfer tools, which require a license to run.
+
+To obtain a FreeSurfer license, simply register for free at
+https://surfer.nmr.mgh.harvard.edu/registration.html.
+
+When using manually-prepared environments or singularity, FreeSurfer will search
```
Comment (josephmje) [inline docs/usage.rst]:
```suggestion
When using manually-prepared environments or Singularity, FreeSurfer will search 
```

File: docs/usage.rst [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+To obtain a FreeSurfer license, simply register for free at
+https://surfer.nmr.mgh.harvard.edu/registration.html.
+
+When using manually-prepared environments or singularity, FreeSurfer will search 
+for a license key file first using the ``$FS_LICENSE`` environment variable and then 
+in the default path to the license key file (``$FREESURFER_HOME/license.txt``). 
+If using the ``--cleanenv`` flag and ``$FS_LICENSE`` is set, use ``--fs-license-file $FS_LICENSE`` 
+to pass the license file location to *dMRIPrep*.
+
+It is possible to run the docker container pointing the image to a local path
```
Comment (josephmje) [inline docs/usage.rst]:
```suggestion
It is possible to run the docker container pointing the image to a local path
```

File: setup.cfg [hunk 7]
```diff
@@ -50,11 +50,14 @@ datalad = datalad
...
     packaging
-    pydot >=1.2.3
+    pydot >= 1.2.3
     pydotplus
-    sphinx >=1.5.3
+    sphinx >= 2.1.2
     sphinx-argparse
     sphinx_rtd_theme
+    sphinxcontrib-apidoc ~= 0.3.0
+    sphinxcontrib-napoleon
```
Comment (josephmje) [inline setup.cfg]: should the versions in `requirements.txt` also be pinned here?

File: docs/usage.rst [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+Usage
+-----
+
+Execution and the BIDS format
+-----------------------------
+The *dMRIPrep* workflow takes as principal input the path of the dataset
+that is to be processed.
+The input dataset is required to be in valid :abbr:`BIDS (Brain Imaging Data
+Structure)` format, and it must include at least one T1w structural image and
+(unless disabled with a flag) a BOLD series.
```
Comment (josephmje) [inline docs/usage.rst]:
```suggestion
(unless disabled with a flag) a dwi series.
```

File: docs/installation.rst [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+------------
+Installation
+------------
+There are two ways of getting *dMRIPrep* installed:
+
+* within a `Manually Prepared Environment (Python 3.5+)`_, also known as
+  *bare-metal installation*; or
+* using container technologies (RECOMMENDED), such as
+  `Docker <https://fmriprep.readthedocs.io/en/stable/docker.html>`__ or
```
[the remark this answers is not in the document]
Comment (oesteban) [reply docs/installation.rst]: Eventually, we will generate these docs - but for now, I guess we need to pull from fMRIPrep for this.

File: docs/installation.rst (hunk 1 above)
Comment (oesteban) [reply docs/installation.rst]: Yup, same as above. I didn't want to copy the whole thing.

File: docs/installation.rst (hunk 2 above)
Comment (oesteban) [reply docs/installation.rst]: Ditto

File: setup.cfg (hunk 7 above)
Comment (oesteban) [reply setup.cfg]: They should not for link-dependencies (i.e., pointing to a github repo and commit directly). If you did then Pypi will not accept the package :(. This is the case of napoleon and versioning.

**Comment by oesteban** [comment]
I'll go ahead and merge since we want to test the actual deployment from master.

**Pull request merged**